metadata/src/validation.rs

```diff
@@ -35,142 +32,154 @@ enum TypeBeingHashed {
 }
 
 /// Hashing function utilized internally.
-fn hash(data: &[u8]) -> [u8; 32] {
+fn hash(data: &[u8]) -> [u8; HASH_LEN] {
     sp_core_hashing::twox_256(data)
 }
 
-/// XOR two hashes together. If we have two pseudorandom hashes, then this will
-/// lead to another pseudorandom value. If there is potentially some pattern to
-/// the hashes we are xoring (eg we might be xoring the same hashes a few times),
-/// prefer `concat_and_hash` to give us stronger pseudorandomness guarantees.
-fn xor(a: [u8; 32], b: [u8; 32]) -> [u8; 32] {
-    let mut out = [0u8; 32];
+/// XOR two hashes together. Only use this when you don't care about the order
+/// of the things you're hashing together.
+fn xor(a: [u8; HASH_LEN], b: [u8; HASH_LEN]) -> [u8; HASH_LEN] {
+    let mut out = [0u8; HASH_LEN];
     for (idx, (a, b)) in a.into_iter().zip(b).enumerate() {
         out[idx] = a ^ b;
     }
     out
 }
 
-/// Combine two hashes or hash-like sets of bytes together into a single hash.
-/// `xor` is OK for one-off combinations of bytes, but if we are merging
-/// potentially identical hashes, this is a safer way to ensure the result is
-/// unique.
-fn concat_and_hash(a: [u8; 32], b: [u8; 32]) -> [u8; 32] {
-    let mut out = [0u8; HASH_LEN * 2];
-    out[0..HASH_LEN].copy_from_slice(&a[..]);
-    out[HASH_LEN..].copy_from_slice(&b[..]);
-    hash(&out)
+// Combine some number of HASH_LEN byte hashes and output a single HASH_LEN
+// byte hash to uniquely represent the inputs.
+macro_rules! count_idents {
+    () => { 0 };
+    ($n:ident $($rest:ident)*) => { 1 + count_idents!($($rest)*) }
+}
+macro_rules! concat_and_hash_n {
+    ($name:ident($($arg:ident)+)) => {
+        fn $name($($arg: &[u8; HASH_LEN]),+) -> [u8; HASH_LEN] {
+            let mut out = [0u8; HASH_LEN * count_idents!($($arg)+)];
+            let mut start = 0;
+            $(
+                out[start..start+HASH_LEN].copy_from_slice(&$arg[..]);
+                #[allow(unused_assignments)]
+                { start += HASH_LEN; }
+            )+
+            hash(&out)
+        }
+    }
 }
+concat_and_hash_n!(concat_and_hash2(a b));
+concat_and_hash_n!(concat_and_hash3(a b c));
+concat_and_hash_n!(concat_and_hash4(a b c d));
+concat_and_hash_n!(concat_and_hash5(a b c d e));
 
 /// Obtain the hash representation of a `scale_info::Field`.
 fn get_field_hash(
     registry: &PortableRegistry,
     field: &Field<PortableForm>,
     visited_ids: &mut HashSet<u32>,
-) -> [u8; 32] {
-    let mut bytes = get_type_hash(registry, field.ty.id, visited_ids);
-
-    // XOR name and field name with the type hash if they exist
-    if let Some(name) = &field.name {
-        bytes = xor(bytes, hash(name.as_bytes()));
-    }
+) -> [u8; HASH_LEN] {
+    let field_name_bytes = match &field.name {
```

Comment:
small nit: 
```rust
let field_name_bytes = field.name.as_ref().map(|name| hash(name.as_bytes())).unwrap_or_default()
```

Reply:
I think the original is easier to read anyways :)